isobus/include/isobus/isobus/isobus_virtual_terminal_objects.hpp

```diff
@@ -76,6 +86,3886 @@ namespace isobus
+		void set_child_y(std::uint16_t index, std::int16_t yOffset);
+
+		/// @brief Offsets all child objects with the specified ID by the amount specified relative to its parent
+		/// @param[in] childObjectID The object ID of the children to offset
+		/// @param[in] xOffset The relative amount to offset the object(s) by in the X axis
+		/// @param[in] yOffset The relative amount to offset the object(s) by in the Y axis
+		/// @returns true if any child matched the specified object ID, otherwise false if no children were found with the specified ID.
+		bool offset_all_children_with_id(std::uint16_t childObjectID, std::int8_t xOffset, std::int8_t yOffset);
+
+		/// @brief Removes an object reference from another object. All fields must exactly match for the object to be removed.
+		/// This is because objects can have multiple of the same child at different places, so we can't infer which one to
+		/// remove without the exact position.
+		/// @param[in] objectIDToRemove The object ID of the child to remove
+		/// @param[in] relativeXLocation The X offset of this object to its parent
+		/// @param[in] relativeYLocation The Y offset of this object to its parent
+		void remove_child(std::uint16_t objectIDToRemove, std::int16_t relativeXLocation, std::int16_t relativeYLocation);
+
+		/// @brief Removes the last added child object.
+		/// This is meant to be a faster way to deal with objects that only have a max of 1 child.
+		void pop_child();
+
+	protected:
+		/// @brief Storage for child object data
+		class ChildObjectData
+		{
+		public:
+			/// @brief Default constructor for child object data with default values
+			ChildObjectData() = default;
+
+			/// @brief Constructor that initializes all members with parameters
+			/// @param[in] objectId The object ID of this child object
+			/// @param[in] x The x location of this child relative to the parent object
+			/// @param[in] y The y location of this child relative to the parent object
+			ChildObjectData(std::uint16_t objectId,
+			                std::int16_t x,
+			                std::int16_t y);
+			std::uint16_t id = NULL_OBJECT_ID; ///< Object identifier. Shall be unique within the object pool.
+			std::int16_t xLocation = 0; ///< Relative X location of the top left corner of the object
+			std::int16_t yLocation = 0; ///< Relative Y location of the top left corner of the object
+		};
+
+		/// @brief A helper function to swap a child object of specified type with a new object.
+		/// This is meant to be used to replace attributes of which there can be only 1, such as a string variable in an output string.
+		/// @param[in] newID The new child object's ID
+		/// @param[in] typeToRemove The object type of both the current child, and the new child (they must match)
+		/// @returns True if the replacement succeeded, otherwise false
+		bool replace_only_child_of_type(std::uint16_t newID, VirtualTerminalObjectType typeToRemove);
```

Comment:
I'm not sure how this can work? Shouldn't you want to simply overwrite the attribute of let's say an output string to achieve this? 

Consider the following (incomplete) pool: 
- Output string (id=20)
- String variable (id=30)
- String variable (id=31)

Then you'd simply want to change the variable reference of the "Output string (id=20)" from "30" to "31"? Instead of replacing things in the object pool like done here?

Reply:
Well, it's not actually replacing an object, per-se, so maybe the name could be clearer... let me explain from the beginning.

So, some of the reason why this exists was to try and reduce the amount of functions and variables in each object. The base object has a `std::vector<ChildObjectData>`, so I've tried to make use of that as much as possible. So if an object like an input number can only ever have 1 of certain kinds of references (number variable, and font attributes in this case) I just put them in that vector instead of making special setters and getters for them. I don't know if this is a good thing or a bad thing, but since I have to do a lot of iterating over objects' children recursively in the VT server, it's maybe more consistent? 

Of course then I still had to support the case where someone wanted to change that child, especially something like a `FontAttributes`, so I made this function to do it for me. It replaces the object ID of the old object in the child vector with the object id of the new child in the vector. So it doesn't actually "replace" something in the pool, if you know what I mean, it's just changing the object ID of a child to another object's ID.

So, I guess I think this function is doing as you said: `change the variable reference of the "Output string (id=20)" from "30" to "31"` it's just a little less than obvious.

If you think it would be clearer to rename it, I could do that. Or if you would prefer concrete child variables with named setters and getters I could do that too, it'll just mean a longer file.

Reply:
Okay hmm I think I get it? Correct me if I'm wrong.

It now uses the `children` list of an object to store objects 'references' as if they were an actual child of the object. But then you don't actually use it as a child, but instead as a reference? I guess this would work yeah, though that also means we have to filter this `children` list when we want to serialize the object. Also a small downside is it requires at least 4 bytes extra storage as `ChildObjectData` also has the locations which are not used when acting as an objectId reference only. IMO not that big of a deal, but will it be worth the hassle?

To be completely honest, I think for _us in the future_, and also other new maintainers, it might be more consistent and clear if we just have for each attribute a dedicated member field of the class. Then no-one has to guess how it's stored and I think it's the most straight forward. 

I do like the way you are thinking of making the file smaller (and hence less code to maintain), but I think there might be too many edge cases attached to this, for it to worth it hmmm... What do you think?

Reply:
To be honest I hadn't really given a lot of thought to serializing the pool yet, so that is a good thing to point out. And yes, there is definitely a 4 byte RAM cost per reference.

I'm not particularly attached to my weird child object thing here, so I'm inclined to agree that maybe it'll be less weird/confusing if we just use dedicated fields for non-child references. I can make those changes

Reply:
Gentleman, 

I'm following the progress of this project, awesome work by the way.
About the VTOP, here goes my take. In the test's I've done here (this is a kind of weekend project for experimentation about the subject), the results using a serialized structure to store all the object in RAM, seem to be working. 
It's not as fast I wished it to be, but perhaps can be of some use for you. 
Also some feedback would be appreciated.
I'm using the following to parse a VTOP and, also, as a base to another class that can draw the elements based on the attributes of each into an LCD screen (actually at this development stage using an ESP32.

[opparser.zip](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/files/13429535/opparser.zip)

PS.: Sorry for not using such a modern C++ implementation, old habits die hard. I'll get there some day.

Reply:
Hey there!

Not to get too off-topic on this PR, but here, take a look at [our parser](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/2cbc71eee5beedb184be0ac4188fba515a4e14f0/isobus/src/isobus_virtual_terminal_server_managed_working_set.cpp#L145) as well for how we've been approaching it. It's maybe a bit heavy-weight for an ESP32 admittedly.

Some feedback I guess would be that it's mostly actually C! So, best-practices aside (such as having `using namespace` in a header file), I worry about how you're essentially just casting the pool data to the different types which would result in very strange things happening if the pool schema is not valid, possibly a hardfault/segfault. There might also be concerns between machines with different memory endianness? Though it's clearly working for ESP32 at least. Casting is efficient, when the pool is properly constructed yes, but intolerant of many OP issues which a server is obligated to report to a client I think unless I am missing something. `appendObject` seems like it could be the slowest part of parsing, lots of memcpy, malloc and free.

I think we are fairly close to making our VT GUI for PC available to the public, so be sure to keep a close eye on things! It has a few issues yet to work out, but I'd say it's maybe 80% complete. I've done some testing with 2 simultaneous working sets each with 10k+ objects and it seemed to hold up fairly well, I think.

Reply:
Removed the function(s) mentioned, and refactored objects so that attributes of concern have dedicated setters and getters